### ISSUE TITLE
File archiving fails when using the same name

### ISSUE THREAD

**Issue by jods4** [NLog/NLog#3608]
**NLog version**: 4.6.3.9764
**Platform**: .Net 4.5

I have NLog configured with file targets + archiving.
My configuration is so that the current file also includes the date, i.e. it has the same name as its future archive. -> This is a requirement because we don't want the file to change when archiving.

In principle it should work as this piece of code checks if the archive target is the same as current log, which results in a no-op:
https://github.com/NLog/NLog/blob/dev/src/NLog/Targets/FileTarget.cs#L1259

Now the bug seems to be that this comparaison uses non-normalized paths.
In my config I have a relative path `..\logs\fabrikam.txt`. 

When looking at NLog internal log, this results in this comparison:
```
Archiving C:\Fabrikam\logs\fabrikam-2019-10.txt to C:\Fabrikam\IIS\..\logs\fabrikam-2019-10.txt
```
As you can see they resolve to the same file, but the strings are different.
So the code continue into `ArchiveFileAppendExisting`, here:
`https://github.com/NLog/NLog/blob/dev/src/NLog/Targets/FileTarget.cs#L1274`
Which fails badly, probably because it tries to append to itself.

I think the paths need to be normalized before they are compared.

We run with `throwExceptions` so this prevents our app from starting.

For reference, here's the internal log:
```
2019-10-01 13:40:56.9385 Debug FileTarget(Name=logfile): Process file 'C:\Fabrikam\IIS\..\logs\Fabrikam.log_2019-10.txt' on startup
2019-10-01 13:40:56.9541 Info FileTarget(Name=logfile): Archive mutex not available: C:\Fabrikam\IIS\..\logs\Fabrikam.log_2019-10.txt
2019-10-01 13:40:56.9541 Debug FileTarget(Name=logfile): Do archive file: 'C:\Fabrikam\IIS\..\logs\Fabrikam.log_2019-10.txt'
2019-10-01 13:40:56.9697 Info FileTarget(Name=logfile): Archiving C:\Fabrikam\logs\Fabrikam.log_2019-10.txt to C:\Fabrikam\IIS\..\logs\Fabrikam.log_2019-10.txt
2019-10-01 13:40:56.9697 Info FileTarget(Name=logfile): Already exists, append to C:\Fabrikam\IIS\..\logs\Fabrikam.log_2019-10.txt
2019-10-01 13:40:56.9697 Warn FileTarget(Name=logfile): Failed to archive file 'C:\Fabrikam\IIS\..\logs\Fabrikam.log_2019-10.txt'. Exception: System.IO.IOException: The process cannot access the file 'C:\Fabrikam\logs\Fabrikam.log_2019-10.txt' because it is being used by another process.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at NLog.Targets.FileTarget.ArchiveFileAppendExisting(String fileName, String archiveFileName)
   at NLog.Targets.FileTarget.ArchiveFile(String fileName, String archiveFileName)
   at NLog.Targets.FileTarget.DoAutoArchive(String fileName, LogEventInfo eventInfo, DateTime previousLogEventTimestamp, Boolean initializedNewFile)
   at NLog.Targets.FileTarget.TryArchiveFile(String fileName, LogEventInfo ev, Int32 upcomingWriteSize, DateTime previousLogEventTimestamp, Boolean initializedNewFile)
```

**Comment by jods4** [comment]
Strangely, I don't have this issue on another service that runs NLog 4.4.5791.0, could this be a recent regression?

**Comment by snakefoot** [comment]
Thank you for reporting this. And excellent job with activating the internal-logger.

Could you also update your original post with the actual FileTarget-configuration?

Btw. you should be very careful with using `throwExceptions="true"`, since it only exists for sake of unit-testing the NLog-library. In production code it can have unwanted side-effects like unhandled exceptions. Recommend using `throwConfigExceptions="true"`

**Comment by jods4** [comment]
@snakefoot I will post the actual config when I'm at work.

We use `throwExceptions` because some of our logs are sensible and we don't want our service to run if the logs are silently failing, e.g. because the file is locked by another process, the folder is gone or its ACL have changed. Think auditing or disaster recovery log.

It's a bit sad that this config applies to _all_ our logs and minor situations such as log archiving, but hey, at least the critical ones are protected. And it catches bugs such as this one.

**Comment by jods4** [comment]
Here's the target that failed:
```xml
    <target name="logfile" xsi:type="File"
            fileName="..\logs\fabrikam.log_${date:format=yyyy-MM}.txt"
            maxArchiveFiles="6"
            archiveNumbering="Date"
            archiveDateFormat="yyyy-MM"
            archiveEvery="Month"
            archiveFileName="..\logs\fabrikam.log_{#}.txt"
            layout="${date:format=dd.MM.yyyy HH\:mm\:ss.fff zzz} ${level:uppercase=true} ${message} ${exception:format:Message,Type,StackTrace:maxInnerExceptionLevel=3:innerFormat=Message,Type,StackTrace:separator=\n}"
            concurrentWrites="True"
            keepFileOpen="True" />
```

**Comment by snakefoot** [comment]
Know that with NLog 4.5 then you can remove all archive parameters except maxArchiveFiles


And it will correctly cleanup. This was not possible with NLog 4.4 and older.


Sent from my Sony Xperia


---- jods wrote ----


Here's the target that failed:

    <target name="logfile" xsi:type="File"
            fileName="..\logs\fabrikam.log_${date:format=yyyy-MM}.txt"
            maxArchiveFiles="6"
            archiveNumbering="Date"
            archiveDateFormat="yyyy-MM"
            archiveEvery="Month"
            archiveFileName="..\logs\fabrikam.log_{#}.txt"
            layout="${date:format=dd.MM.yyyy HH\:mm\:ss.fff zzz} ${level:uppercase=true} ${message} ${exception:format:Message,Type,StackTrace:maxInnerExceptionLevel=3:innerFormat=Message,Type,StackTrace:separator=\n}"
            concurrentWrites="True"
            keepFileOpen="True" />

**Comment by jods4** [comment]
Thanks for the tip! 
I'm curious how does it work? 
Does it work only because I added the date in `fileName` + I'm using the same folder for archiving?

**Comment by snakefoot** [comment]
Yes it is for same folder archive. NLog 4.5 improves how it generates a file wildcard from the fileName. Which is used for cleanup.


Sent from my Sony Xperia


---- jods wrote ----


Thanks for the tip!
I'm curious how does it work?
Does it work only because I added the date in fileName + I'm using the same folder for archiving?

**Comment by 304NotModified** [comment]
fixed by #3609
